core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java

```diff
@@ -145,22 +144,24 @@ private NodePartitionMap getNodePartitioningMap(
             Session session,
             PartitioningHandle partitioningHandle,
             Map<Integer, List<InternalNode>> bucketToNodeCache,
-            AtomicReference<List<InternalNode>> systemPartitioningCache)
+            AtomicReference<List<InternalNode>> systemPartitioningCache,
+            int hashPartitionCount)
     {
         requireNonNull(session, "session is null");
         requireNonNull(partitioningHandle, "partitioningHandle is null");
 
         if (partitioningHandle.getConnectorHandle() instanceof SystemPartitioningHandle) {
-            return systemNodePartitionMap(session, partitioningHandle, systemPartitioningCache);
+            return systemNodePartitionMap(session, partitioningHandle, systemPartitioningCache, hashPartitionCount);
         }
 
         if (partitioningHandle.getConnectorHandle() instanceof MergePartitioningHandle mergeHandle) {
-            return mergeHandle.getNodePartitioningMap(handle -> getNodePartitioningMap(session, handle, bucketToNodeCache, systemPartitioningCache));
+            return mergeHandle.getNodePartitioningMap(handle ->
+                    getNodePartitioningMap(session, handle, bucketToNodeCache, systemPartitioningCache, hashPartitionCount));
         }
 
         Optional<ConnectorBucketNodeMap> optionalMap = getConnectorBucketNodeMap(session, partitioningHandle);
         if (optionalMap.isEmpty()) {
-            return systemNodePartitionMap(session, FIXED_HASH_DISTRIBUTION, systemPartitioningCache);
+            return systemNodePartitionMap(session, FIXED_HASH_DISTRIBUTION, systemPartitioningCache, hashPartitionCount);
```

Comment:
I would say we could ignore this case for now and just use `getMaxHashPartitionCount`. I think it's for insert path (merge) anyway.

Alternatively, if we had `Optional<Int> hashPartitionCount` in `PartitioningHandle` then we could extract it from there.

However, as mentioned in other comment I'm inclining that we should move `Optional<Int> hashPartitionCount` to `SystemPartitioningHandle` for now.